File: data/EXP01/python/2497/2497.output_3.py
# @lc app=leetcode id=2497 lang=python3
from typing import List

class Solution:
    def maxStarSum(self, vals: List[int], edges: List[List[int]], k: int) -> int:
        n = len(vals)
        adj_list = [[] for _ in range(n)]  # adjacency list representation of the graph
        for edge in edges:
            u, v = edge
            adj_list[u].append(v)
            adj_list[v].append(u)
        
        max_sum = float('-inf')
        
        def dfs(node, parent):
            nonlocal max_sum
            
            sum_with_node = vals[node]
            num_edges = 0
            
            for neighbor in adj_list[node]:
                if neighbor == parent:
                    continue
                sum_with_neighbor, num_edges_with_neighbor = dfs(neighbor, node)
                sum_with_node += sum_with_neighbor
                num_edges += num_edges_with_neighbor + 1
            
            if num_edges <= k:
                max_sum = max(max_sum, sum_with_node)
            
            return sum_with_node, num_edges
        
        dfs(0, -1)  # Start the DFS from node 0
        
        return max_sum

File: data/EXP01/golang_no_system_var/2497/2497.output_3.go
func maxStarSum(vals []int, edges [][]int, k int) int {
    n := len(vals)
    graph := make(map[int][]int)
    for _, edge := range edges {
        a, b := edge[0], edge[1]
        graph[a] = append(graph[a], b)
        graph[b] = append(graph[b], a)
    }
    
    maxSum := vals[0] // initialize maxSum with the value of the first node
    
    var dfs func(int, int, int) int
    dfs = func(node, parent, edgesUsed int) int {
        sum := vals[node]
        children := 0
        
        for _, neighbor := range graph[node] {
            if neighbor != parent {
                childSum := dfs(neighbor, node, edgesUsed+1)
                if childSum > 0 {
                    sum += childSum
                    children++
                }
            }
        }
        
        if children == 0 && edgesUsed <= k {
            return sum
        }
        
        if edgesUsed < k && sum > maxSum {
            maxSum = sum
        }
        
        return sum
    }
    
    dfs(0, -1, 0) // start DFS from the first node
    
    return maxSum
}
